Forge.TreeWalker/src/TreeWalkerParameters.cs

```diff
@@ -180,8 +188,89 @@ public class TreeWalkerParameters
+        /// <param name="token">The cancellation token.</param>
+        public TreeWalkerParameters(
+            Guid sessionId,
+            ForgeTree forgeTree,
+            IForgeDictionary forgeState,
+            ITreeWalkerCallbacksV2 callbacksV2,
+            CancellationToken token)
+        {
+            if (sessionId == Guid.Empty) throw new ArgumentNullException("sessionId");
+            if (forgeTree == null) throw new ArgumentNullException("forgeTree");
+            if (forgeState == null) throw new ArgumentNullException("forgeState");
+            if (callbacksV2 == null) throw new ArgumentNullException("callbacksV2");
+            if (token == null) throw new ArgumentNullException("token");
+
+            this.SessionId = sessionId;
+            this.ForgeTree = forgeTree;
+            this.ForgeState = forgeState;
+            this.CallbacksV2 = callbacksV2;
+            this.Token = token;
+        }
+
+        /// <summary>
+        /// Instantiates a TreeWalkerParameters object with the properties that are required to instantiate a TreeWalkerSession object.
+        /// Note: Recommended to cache the deserialized ForgeTree in your app and use that to avoid Forge deserializing the JsonSchema each initialize.
+        /// </summary>
+        /// <param name="sessionId">The unique identifier for this session.</param>
+        /// <param name="jsonSchema">The JSON schema.</param>
+        /// <param name="forgeState">The Forge state.</param>
+        /// <param name="callbacksV2">The callbacks object.</param>
+        /// <param name="token">The cancellation token.</param>
+        public TreeWalkerParameters(
+            Guid sessionId,
+            string jsonSchema,
+            IForgeDictionary forgeState,
+            ITreeWalkerCallbacksV2 callbacksV2,
+            CancellationToken token)
+        {
+            if (sessionId == Guid.Empty) throw new ArgumentNullException("sessionId");
+            if (string.IsNullOrWhiteSpace(jsonSchema)) throw new ArgumentNullException("jsonSchema");
+            if (forgeState == null) throw new ArgumentNullException("forgeState");
+            if (callbacksV2 == null) throw new ArgumentNullException("callbacksV2");
+            if (token == null) throw new ArgumentNullException("token");
+
+            this.SessionId = sessionId;
+            this.JsonSchema = jsonSchema;
+            this.ForgeState = forgeState;
+            this.CallbacksV2 = callbacksV2;
+            this.Token = token;
+        }
+        #endregion Constructor with ITreeWalkerCallbacksV2
+
+        private void TryUpgradeToITreeWalkerCallbacksV2(ITreeWalkerCallbacks callbacks)
+        {
+            if (callbacks is ITreeWalkerCallbacksV2)
+            {
+                ITreeWalkerCallbacksV2 callbacksV2 = callbacks as ITreeWalkerCallbacksV2;
+                if (callbacksV2 != null)
+                {
+                    this.CallbacksV2 = callbacksV2;
```

Comment:
>this.CallbacksV2 = callbacksV2; [](start = 20, length = 31)

Interesting idea to allow passing in ITreeWalkerCallbacks and attempting to convert to V2 on the users behalf. Since we are doing this, perhaps we don't even need the new constructors?

Reply:
Technically, yes, with this change, we don't need the new constructors. However, we are encouraging users to use the new constructors if possible. One we confirm there is no user for the old constructor, we can remove them.

Reply:
Technically, yes, with this change, we don't need the new constructors. However, we are encouraging users to use the new constructors if possible. One we confirm there is no user for the old constructor, we can remove them.

---
In reply to: [504157718](https://github.com/microsoft/Forge/pull/32#discussion_r504157718) [](ancestors = 504157718)